src/server/game/Config/GameConfig.cpp

```diff
@@ -0,0 +1,210 @@
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "DatabaseEnv.h"
+#include "GameConfig.h"
+#include "Log.h"
+#include "Timer.h"
+#include "Util.h"
+
+GameConfig* GameConfig::instance()
+{
+    static GameConfig instance;
+    return &instance;
+}
+
+void GameConfig::CleanAll()
+{
+    _boolOptions.clear();
+    _intOptions.clear();
+    _floatOptions.clear();
+    _stringOptions.clear();
+}
+
+void GameConfig::AddBoolOption(std::string const& optionName, bool value /*= false*/)
+{
+    auto const& itr = _boolOptions.find(optionName);
+    if (itr != _boolOptions.end())
+    {
+        TC_LOG_FATAL("server.loading", "> Bool option (%s) is exist!", optionName.c_str());
+        return;
+    }
+
+    _boolOptions.insert(std::make_pair(optionName, value));
+}
+
+void GameConfig::AddIntOption(std::string const& optionName, int32 value /*= 0*/)
+{
+    auto const& itr = _intOptions.find(optionName);
+    if (itr != _intOptions.end())
+    {
+        TC_LOG_FATAL("server.loading", "> Int option (%s) is exist!", optionName.c_str());
+        return;
+    }
+
+    _intOptions.insert(std::make_pair(optionName, value));
+}
+
+void GameConfig::AddFloatOption(std::string const& optionName, float value /*= 1.0f*/)
+{
+    auto const& itr = _floatOptions.find(optionName);
+    if (itr != _floatOptions.end())
+    {
+        TC_LOG_FATAL("server.loading", "> Float option (%s) is exist!", optionName.c_str());
+        return;
+    }
+
+    _floatOptions.insert(std::make_pair(optionName, value));
+}
+
+void GameConfig::AddStringOption(std::string const& optionName, std::string const& value /*= ""*/)
+{
+    auto const& itr = _stringOptions.find(optionName);
+    if (itr != _stringOptions.end())
+    {
+        TC_LOG_FATAL("server.loading", "> Int option (%s) is exist!", optionName.c_str());
+        return;
+    }
+
+    _stringOptions.insert(std::make_pair(optionName, value));
+}
+
+void GameConfig::AddOption(std::string const& optionName, GameConfigType type, std::string const& value, std::string const& defaultValue)
+{
+    switch (type)
+    {
+    case GameConfigType::GAME_CONFIG_TYPE_BOOL:
+        AddBoolOption(optionName, value.empty() ? StringToBool(defaultValue) : StringToBool(value));
+        break;
+    case GameConfigType::GAME_CONFIG_TYPE_INT:
+        AddIntOption(optionName, value.empty() ? std::stoi(defaultValue) : std::stoi(value));
+        break;
+    case GameConfigType::GAME_CONFIG_TYPE_FLOAT:
+        AddFloatOption(optionName, value.empty() ? std::stof(defaultValue) : std::stof(value));
+        break;
+    case GameConfigType::GAME_CONFIG_TYPE_STRING:
+        AddStringOption(optionName, value.empty() ? defaultValue : value);
+        break;
+    default:
+        ABORT();
```

Comment:
does this just crash worldserver with not useful message about why it did that ?

Reply:
Hm, i agree